File: archive/2020/solutions/SealSharing.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

const int MAXN = 105;

const int MOD = 1e9 + 7;

ll a[MAXN];
bool vis[MAXN];

map<int, vector<int>> change;
vector<int> need;
vector<vector<int>> store;
map<int, int> mapUpdates;

int main(){
    int n; cin >> n;
    for(int i = 0; i<n; i++) cin >> a[i];

    int q; cin >> q;
    int maxt = 0;
    vector<pair<int, int>> updates;
    for(int i = 0; i<q; i++){
        int k, t; cin >> k >> t;
        maxt = max(maxt, t);
        updates.emplace_back(make_pair(k, t));
    }

    set<int> tmp;
    for(auto x: updates) tmp.insert(x.second);
    for(auto x: tmp) need.emplace_back(x);
    for(int i = 0; i<need.size(); i++) mapUpdates[need[i]] = i;
    vector<int> cur;
    for(int i = 0; i<n; i++) cur.emplace_back(0);
    cur[0] = 1;
    int times = 0;
    for(int x: need){
        while(times<x){
            vector<int> nxt;
            nxt.resize(n);
            for(int i = 0; i<n; i++) nxt[i] = cur[i] + cur[(i-1+n)%n];
            for(int i = 0; i<n; i++) if(nxt[i]>=MOD) nxt[i] -= MOD;
            swap(nxt, cur);
            times++;
        }
        store.emplace_back(cur);
    }

    for(auto u: updates){
        int k = u.first;
        int t = mapUpdates[u.second];
        for(int i = 0; i<n; i++) vis[i] = false;
        vector<ll> nxt;
        nxt.resize(n);
        vector<ll> use;

        for(int i = 0; i<n; i++){
            if(vis[i]) continue;
            vector<int> v;
            int cval = i;
            while(!vis[cval]){
                vis[cval] = true;
                v.emplace_back(cval);
                cval += k;
                if(cval>=n) cval-=n;
            }

            if(use.size() == 0){
                use.resize(v.size());
                for(int j = 0; j<n; j++){
                    int idx = j%use.size();
                    use[idx] += store[t][j];
                    if(use[idx]>=MOD) use[idx] -= MOD;
                }
            }

            for(int j = 0; j<v.size(); j++){
                for(int l = 0; l<v.size(); l++){
                    int idx = (j + l)%v.size();
                    nxt[v[idx]] += (use[l]*a[v[j]])%MOD;
                    if(nxt[v[idx]]>=MOD) nxt[v[idx]] -= MOD;
                }
            }
        }

        for(int i = 0; i<n; i++) a[i] = nxt[i];
    }

    for(int i = 0; i<n; i++) cout << a[i] << " ";
    cout << endl;
}

File: archive/2020/solutions/SealSharing.py
def main():
    MOD = 1000000007;
    
    a = [None] * 105
    vis = [0] * 105

    change = {}
    mapUpdates = {}
    store = []
    

    n = int(input())
    a = list(map(int, input().split()))

    q = int(input())
    maxt = 0
    updates = []
    for i in range(q):
        x = input().split()
        k = int(x[0])
        t = int(x[1])
        maxt = max(maxt, t)
        updates.append(k)
        updates.append(t)

    need = []
    cur = [0] * n
    cur[0] = 1

    tmp = set()
    for i in range(q):
        tmp.add(updates[2 * i + 1])
    for i in tmp:
        need.append(i)
    need.sort()
    for i in range(len(need)):
        mapUpdates[need[i]] = i
    times = 0
    for x in need:
        while times < x:
            nxt = [None] * n
            for i in range(n):
                nxt[i] = cur[i] + cur[(i-1+n)%n]
            for i in range(n):
                if (nxt[i] >= MOD): nxt[i] -= MOD
                cur[i] = nxt[i]
            times += 1
    
        store.append(cur.copy())

    for i in range(q):
        k = updates[2 * i]
        t = mapUpdates[updates[2 * i + 1]]
        nxt = [0] * n
        use = []
        vis = [0]*n #
        
        for i in range(n):
            if vis[i]:
                continue
            v = []
            cval = i
            while (not vis[cval]):
                vis[cval] = 1
                v.append(cval)
                cval += k
                if cval >= n: cval -= n

            if len(use) == 0:
                use = [0] * len(v)
                for j in range(n):
                    idx = j % len(use)
                    use[idx] += store[t][j]
                    if use[idx] >= MOD: use[idx] -= MOD

            for j in range(len(v)):
                for l in range(len(v)):
                    idx = (j + l) % len(v)
                    nxt[v[idx]] += (use[l] * a[v[j]]) % MOD
                    nxt[v[idx]] %= MOD

        for i in range(n):
            a[i] = nxt[i]

    for i in range(n):
        print(a[i], end = " ")
    print()

main()
